### PR TITLE
Add support for toggling Fortran features

### LINKED ISSUE #363: Reclaiming `.f` file extension for modern Fortran (free form)
This was first seriously suggested here:

https://github.com/fortran-lang/fpm/issues/359#issuecomment-779377262

This deserves an issue of its own. Let's discuss the pros and cons, here is a start:

### Cons

* Change of established (by now) conventions to use `.f` for fixed-form and `.f90` for free-form (see @sblionel's [post](https://stevelionel.com/drfortran/2013/01/11/doctor-fortran-in-source-form-just-wants-to-be-free/) for the history of this). Thus creating even more confusion.
* Syntax highlighting in Vim, GitHub, etc. might be incorrect for free-form code in `.f` files
* Some existing Fortran users might be disappointed with this choice
* While this can be implemented in `fpm` easily, it might break some other existing tools around Fortran

### Pros

* `.f` is a nice extension, natural for Fortran
* It does not have 90 in the name, thus nothing "old" about it
* Writing fixed-form is not recommended for new projects, so the default good extension should be reserved for free-form, rather than fixed-form
* New Fortran users might like this new extension more

Relevant other issues: #250, #359.

### LINKED ISSUE #577: Support disabling of implicit typing in package manifest
### Motivation

Allow packages to opt-out from default implicit typing in Fortran. This allows to get the full befits of `implicit none` in every scope without having to remember typing it.

More difficult is whether we can find options in every compiler to actually guarantee that this is working.

### Specification

The *build* table seems most appropriate for this option:

```toml
[build]
implicit-typing = false
```

### Prior Art

Compiler flags:

| compiler | option |
| --- | --- |
| GFortran | `-fimplicit-none` |
| NVHPC | `-Mdclchk` |
| nagfor | `-u` |
| Cray/HPE | `-eI` |
|SilverFrost FTN95 | `/IMPLICIT_NONE` | 
| ifort (Windows) | `/warn:declarations,errors` |
| ifort (Linux, macOS) | `-warn declarations,errors` |
| flang | `-fimplicit-none-type-always` |
| IBM XL Fortran | `-qundef` or `-u` |

### Additional Information

Related issue https://github.com/fortran-lang/fpm/issues/359
Draft PR: https://github.com/fortran-lang/fpm/pull/506

### PR DESCRIPTION
**Pull request by awvwgk**
This has been hanging around stale long enough, here is a draft implementation for disabling implicit typing and external interfaces by default with opt-in toggles on per package basis. Also allows to switch globally between fixed and free format for source files.

- [x] *implicit-typing*: toggle default implicit typing rules
  - option to disable in GFortran
  - option to enable in LFortran
- [x] *implicit-external*: toggle implicit external interfaces
  - option to disable in GFortran
  - option to enable in LFortran
- [x] *source-form*: select source form ("free", "fixed", or "default")
  -  option to set free/fixed form in GFortran
  - option to set fixed form in LFortran

*Usage*

```toml
[fortran]
implicit-typing = true  # default: false
implicit-external = true  # default: false
source-form = "fixed"  # default: "free"
```

Documentation at https://github.com/fortran-lang/fpm-docs/pull/98
Closes #577 
Closes #363

### CODE REVIEW

File: src/fpm_compiler.F90 [hunk 1]
```diff
@@ -137,10 +139,14 @@ module fpm_compiler
...
-    flag_gnu_warn = " -Wall -Wextra -Wimplicit-interface", &
+    flag_gnu_warn = " -Wall -Wextra", &
     flag_gnu_check = " -fcheck=bounds -fcheck=array-temps", &
     flag_gnu_limit = " -fmax-errors=1", &
-    flag_gnu_external = " -Wimplicit-interface"
+    flag_gnu_external = " -Wimplicit-interface", &
+    flag_gnu_no_implicit_typing = " -fimplicit-none", &
+    flag_gnu_no_implicit_external = " -Werror=implicit-interface", &
+    flag_gnu_free_format = " -ffree-form", &
+    flag_gnu_fixed_format = " -ffixed-form"
```
Comment (gnikit) [inline src/fpm_compiler.F90:149]: I would add the other compiler options too (sorry couldn't think of another way of doing it through the _patch suggestion_


```suggestion
    flag_gnu_fixed_format = " -ffixed-form"

character(*), parameter :: &
    flag_pgi_backslash = " -Mbackslash", &
    flag_pgi_traceback = " -traceback", &
    flag_pgi_debug = " -g", &
    flag_pgi_check = " -Mbounds -Mchkptr -Mchkstk", &
    flag_pgi_warn = " -Minform=inform", &
    flag_pgi_free_format = " -Mfree", &
    flag_pgi_fixed_format = " -Mfixed"

character(*), parameter :: &
    flag_ibmxl_backslash = " -qnoescape"

character(*), parameter :: &
    flag_intel_backtrace = " -traceback", &
    flag_intel_warn = " -warn all", &
    flag_intel_check = " -check all", &
    flag_intel_debug = " -O0 -g", &
    flag_intel_fp = " -fp-model precise -pc64", &
    flag_intel_align = " -align all", &
    flag_intel_limit = " -error-limit 1", &
    flag_intel_pthread = " -reentrancy threaded", &
    flag_intel_nogen = " -nogen-interfaces", &
    flag_intel_byterecl = " -assume byterecl", &
    flag_intel_free_format = " -free", &
    flag_intel_fixed_format = " -fixed"

character(*), parameter :: &
    flag_intel_backtrace_win = " /traceback", &
    flag_intel_warn_win = " /warn:all", &
    flag_intel_check_win = " /check:all", &
    flag_intel_debug_win = " /Od /Z7", &
    flag_intel_fp_win = " /fp:precise", &
    flag_intel_align_win = " /align:all", &
    flag_intel_limit_win = " /error-limit:1", &
    flag_intel_pthread_win = " /reentrancy:threaded", &
    flag_intel_nogen_win = " /nogen-interfaces", &
    flag_intel_byterecl_win = " /assume:byterecl", &
    flag_intel_free_format_win = " /free", &
    flag_intel_fixed_format_win = " /fixed"

character(*), parameter :: &
    flag_nag_coarray = " -coarray=single", &
    flag_nag_pic = " -PIC", &
    flag_nag_check = " -C", &
    flag_nag_debug = " -g -O0", &
    flag_nag_opt = " -O4", &
    flag_nag_backtrace = " -gline", &
    flag_nag_free_format = " -free", &
    flag_nag_fixed_format = " -fixed"

```

File: src/fpm_compiler.F90 [hunk 2]
```diff
@@ -539,6 +548,67 @@ function get_module_flag(self, path) result(flags)
...
+       case(id_lfortran)
+           flags = flag_lfortran_implicit_external
+
+       end select
+
+    case("free-format")
+       select case(self%id)
+       case(id_caf, id_gcc, id_f95)
+           flags = flag_gnu_free_format
+
```
Comment (gnikit) [inline src/fpm_compiler.F90:590]: Adds free-form compiler flags. I think the grouped ids per `case` are correct but please double check

```suggestion
       case(id_caf, id_gcc, id_f95, id_flang_new)
           flags = flag_gnu_free_format

        case(id_intel_classic_nix, id_intel_llvm_nix, id_intel_classic_mac)
            flags = flag_intel_free_format

        case(id_intel_classic_windows, id_intel_llvm_windows)
            flags = flag_intel_free_format_win

        case(id_pgi, id_nvhpc, id_flang, id_f18)
            flags = flag_pgi_free_format

        case(id_nag)
            flags = flag_nag_free_format

```

File: src/fpm_compiler.F90 [hunk 3]
```diff
@@ -539,6 +548,67 @@ function get_module_flag(self, path) result(flags)
...
+       end select
+
+    case("fixed-format")
+       select case(self%id)
+       case(id_caf, id_gcc, id_f95)
+           flags = flag_gnu_fixed_format
+
```
Comment (gnikit) [inline src/fpm_compiler.F90:597]: Adds fixed-form compile flags (also fixes indentation on the fist couple of lines)

```suggestion
        select case(self%id)
        case(id_caf, id_gcc, id_f95, id_flang_new)
           flags = flag_gnu_fixed_format

        case(id_intel_classic_nix, id_intel_llvm_nix, id_intel_classic_mac)
            flags = flag_intel_fixed_format

        case(id_intel_classic_windows, id_intel_llvm_windows)
            flags = flag_intel_fixed_format_win

        case(id_pgi, id_nvhpc, id_flang, id_f18)
            flags = flag_pgi_fixed_format

        case(id_nag)
            flags = flag_nag_fixed_format

```

**Review by gnikit — CHANGES_REQUESTED** [review]
Very nice you beat me to it (by many, many hours). I left some comments/patch suggestions for adding support for the other compilers (at least for the fixed/free flag). Have a look and let me know if they are correct, if not you can dismiss them.

**Comment by awvwgk** [comment]
Giannis @gnikit, feel free to commit the changes directly to my branch.

**Comment by gnikit** [comment]
> Giannis @gnikit, feel free to commit the changes directly to my branch.

Will do this today

**Comment by milancurcic** [comment]
I think this is a great change. 

The TOML key should be `source-form`, not `source-format`. Likewise, use "fixed-form" and "free-form" instead of "fixed-format" and "free-format" in example names.

`"default"` for the `source-form` makes sense to me--it reads like "use the compiler default, whatever that is".

**Comment by awvwgk** [comment]
Thanks everyone for reviewing and providing feedback. While not all compilers currently support these features, we can add those in the future as we discover the correct options. I'll go ahead and merge now.

**Pull request merged**